packages/core/contracts/oracle/implementation/cross-chain/BeaconOracle.sol

```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.6.0;
+pragma experimental ABIEncoderV2;
+
+import "../../interfaces/OracleAncillaryInterface.sol";
+import "../../../common/implementation/MultiRole.sol";
+import "../../interfaces/FinderInterface.sol";
+import "./BridgeInterface.sol";
+import "../Constants.sol";
+
+/**
+ * @title Simple implementation of the OracleInterface used to communicate price request data cross-chain between
+ * EVM networks. Can be extended either into a "Source" or "Sink" oracle. The intention is that the "Source" Oracle
+ * is the originator of price resolution data, secured by the DVM, and the "Sink" Oracle can both receive this price
+ * resolution data and request prices cross-chain to the "Source" Oracle. The "Sink" is designed to be deployed on
+ * non-Mainnet networks and the "Source" should be deployed on Mainnet.
+ */
+abstract contract BeaconOracle is OracleAncillaryInterface, MultiRole {
+    enum Roles {
+        Owner, // The owner manages the other roles.
+        Requester, // Can make price requests.
```

Comment:
On the side-chain, wouldn't we be pulling this from the registry (i.e. wouldn't it be all registered contracts)? Or are we planning a different permissioning architecture?

Reply:
Let's go with Registry for requests, however do you re-use the same list for price publishers? If not, then do you create another role?

Reply:
Yeah, not sure. Let me explain how I see it, and let me know if you agree:
- Sidechain publisher: bridge.
- Sidechain requester: registry approved contracts.
- Main chain publisher: none? It seems like _this_ contract should grab the price directly from the DVM?
- Main chain requester: bridge.

A few of these don't really fit into multirole, but some do. How do you think we should manage that? Maybe we should just have these methods be non-public in the parent contract and have the derived contracts manage the permissioning individually?

Thoughts?

Reply:
We can also make all publish and request methods callable only by registered contracts, and set the Handler (i.e. the Bridge in your example, but its technically the Handler) as a registered contract.

Reply:
But I do agree with your categorization except that `bridge` should be `generic handler`.

Reply:
Yeah, agreed.